### PR TITLE
Support older FG3 game versions

### PR DESCRIPTION
**Pull request by MsRandom**
This redoes some of what #86 did, specifically the parts supporting legacy FG3 versions.
The gist of what this does is:

- Checking for more userdev config options, specifically `notchObf`, `ats`, and `universalFilters` since they are needed for legacy to work
- Changing the step order for patching the Jars if `notchObf` is true
- Changing the userdev classifier for versions lower than 1.13
- Only merging SRG mappings with official mappings on versions that support them
- Supporting the `mcp_to_srg` template argument option, and also adding unmapped class names to the SRG mappings file as they are needed in older versions.
- Not adding the patch provider unless it is required(and failing if single Jar is used with versions that had a different installer format)

### CODE REVIEW

**Comment by wagyourtail** [comment]
besides launch configs not working for me, this works pretty well. I was able to successfully use it with legacy yarn to backport my mod to 1.12.2

idk if it's possible to fix launching to be compatible for newer java versions maybe by using OF's launchwrapper version 2.3, but that might be outside scope for this pr

**Comment by MsRandom** [comment]
The launch configs can work if you just manually change the JDK version they use, I'll look into fixing them for newer JDK versions though.

File: src/main/java/net/fabricmc/loom/configuration/providers/forge/ForgeProvider.java
Comment (shedaniel) [inline src/main/java/net/fabricmc/loom/configuration/providers/forge/ForgeProvider.java]: Is this using the java internal version parser? If so, please don't

File: src/main/java/net/fabricmc/loom/configuration/providers/mappings/MappingsProviderImpl.java
Comment (shedaniel) [inline src/main/java/net/fabricmc/loom/configuration/providers/mappings/MappingsProviderImpl.java]: Maybe only do this for legacy versions

File: src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java
Comment (shedaniel) [inline src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java]: use replace existing instead of this

File: src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java
Comment (shedaniel) [inline src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java:441]: is this the same behaviour as the previous code?

**Comment by shedaniel** [comment]
> The launch configs can work if you just manually change the JDK version they use, I'll look into fixing them for newer JDK versions though.

That would be out of scope for this PR, getting it working with the original java is sufficient

File: src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java
Comment (MsRandom) [reply src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java:441]: The only difference is now it'd be getting the access transformer from the userdev jar instead of the universal, but it should be the same behavior wise.

File: src/main/java/net/fabricmc/loom/configuration/providers/forge/ForgeProvider.java
Comment (MsRandom) [reply src/main/java/net/fabricmc/loom/configuration/providers/forge/ForgeProvider.java]: `java.lang.Runtime.Version` appears to be public API since Java 9 as far as I can tell.

Comment (Juuxel) [reply src/main/java/net/fabricmc/loom/configuration/providers/forge/ForgeProvider.java]: It's intended for versions of the Java runtime and will break since Minecraft versions don't conform to that pattern. Please use something else.

File: src/main/java/net/fabricmc/loom/configuration/providers/forge/minecraft/MergedForgeMinecraftProvider.java
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/configuration/providers/forge/minecraft/MergedForgeMinecraftProvider.java:72]: But... even merged MC does need it on any semi-recent versions, though?

File: src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java]: Don't hardcode `Type.MERGED` but instead use the field like in the previous code, this'll break single jar MC provider

File: src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java]: This should use the renaming tool from the MCPConfig json, or since it doesn't need to produce specific bytecode anymore, just Tiny Remapper.

File: src/main/java/net/fabricmc/loom/configuration/providers/mappings/MappingsProviderImpl.java
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/configuration/providers/mappings/MappingsProviderImpl.java]:
```suggestion
```

File: src/main/java/net/fabricmc/loom/configuration/providers/forge/ForgeProvider.java
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/configuration/providers/forge/ForgeProvider.java]: Minecraft versions aren't just dot-separated lists of ints so this will also probably break with something.

File: src/main/java/net/fabricmc/loom/configuration/providers/forge/ForgeProvider.java
Comment (MsRandom) [reply src/main/java/net/fabricmc/loom/configuration/providers/forge/ForgeProvider.java]: I made it manually compare the versions. Although it'll still fail on snapshot versions (Should I use `MinecraftVersion` from srgutils instead? that'd have accurate comparing even with snapshots or anything else)

File: src/main/java/net/fabricmc/loom/configuration/providers/forge/ForgeUserdevProvider.java
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/configuration/providers/forge/ForgeUserdevProvider.java]: 1. Those are the default open options for `Files.write` (with the exception of `WRITE` which I think is actually also *needed* here) so they can be left out
2. If there are multiple files, this will just replace the contents each time it loops. `StandardOpenOption.APPEND` is probably better

File: src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java]: A better name is needed here, `minecraftPatchedJar` and `minecraftBinPatchedJar` mean the same thing

File: src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java:369]: Have you checked that the AT is still found as expected on all versions since the updated code isn't searching in all of these locations?

File: src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java:501]: Why was this logic moved here? It works just fine at the end.

File: src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java]:
```suggestion
	private final class UserdevFilter implements OutputConsumerPath.ResourceRemapper {
```

**Review by Juuxel — CHANGES_REQUESTED** [review]

File: src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java
Comment (MsRandom) [reply src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java]: That was an accident, I'll fix it.

File: src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java
Comment (MsRandom) [reply src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java:369]: I did. Using the userdev one works on every version I've tested.

File: src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java
Comment (MsRandom) [reply src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java:501]: It breaks on older versions as the hierarchy for `ShapedRecipe` relies on the forge class `IShapedRecipe`, so remapping doesn't produce correct results since there's a missing class in between the base recipe and shaped recipe classes.

File: src/main/java/net/fabricmc/loom/configuration/providers/forge/minecraft/MergedForgeMinecraftProvider.java
Comment (MsRandom) [reply src/main/java/net/fabricmc/loom/configuration/providers/forge/minecraft/MergedForgeMinecraftProvider.java:72]: It only gives the installer client and server patches, which are not used when merged is used (since MERGED just uses the userdev joined patches)

File: src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java
Comment (MsRandom) [reply src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java]: There's no simple way to fetch the rename tool from the config without having a custom execution context to resolve the arguments, should I do that?

Also, the mappings provider is not setup by that point, so I can't get the tiny mappings with srg to use with Tiny Remapper.

Comment (Juuxel) [reply src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java]: Hmm I think this might be fine then. I'd move the dep string into `Constants.Dependencies` and add a note that it's for remapping on <1.13 Forge.

**Comment by wagyourtail** [comment]
> The launch configs can work if you just manually change the JDK version they use, I'll look into fixing them for newer JDK versions though.

Tried that. Idk, maybe it was another issue

**Comment by MsRandom** [comment]
Is there anything else I should do for this?

**Comment by triphora** [comment]
Really just a matter of waiting for people to be available to review/merge it. And maybe a bit of bugging on Discord would help :laughing:

**Comment by Juuxel** [comment]
I can review after I'm done with my mod updates 🙂

**Comment by hYdos** [comment]
big mod updates... :tiny_potato:

**Comment by SokyranTheDragon** [comment]
I've tried using those to try and do a bit more maintenance on my 1.12 mod. It works great. Except for 2 issues while using MCP mappings. I've noticed:

- Using MCP 1.15 (or older) mappings fails, as they seem to be provided as a .zip file (from Forge Maven, if I'm not mistaken), and not .jar (for example: `mappings "de.oceanlabs.mcp:mcp_stable:39-1.12"`)
- At least with the 1.12 mappings (stable 39), there seems to be some problem with applying them? As in, a wrong one being applied. I've noticed 1 very obvious example of that happening: `Block.getCollisionBoundingBox(IBlockState blockState, IBlockAccess worldIn, BlockPos pos)` somehow got mapped as `Block.getItem(IBlockState worldIn, IBlockAccess pos, BlockPos state)`.

For the first one, I did a quick and hacky workaround using what after a quick search. Mind you - I've got no experience with Gradle plugins, and I'm pretty bad at Gradle itself, so there's definitely a much better way to handle this.
At the end of `DependencyInfo` constructor I've added this:
```java
sourceConfiguration.resolutionStrategy(strategy ->
		strategy.dependencySubstitution(substitutions ->
				substitutions.all(depSubst ->
						depSubst.artifactSelection(selDetails ->
								selDetails.selectArtifact("zip", null, null)))));
```

As for the second one... Well, it's beyond me why it happens. Maybe I'm doing something wrong? I honestly have no idea. I've tried to use older snapshots for 1.12, but I've been getting the same outcome. Maybe it's something with the SRG mapping? Do I need to apply them manually? They seem to be .zip files as well, so maybe the same issue as with MCP mappings? I can provide more info/my build.gradle/etc. if needed.

**Comment by hYdos** [comment]
I'm 90% sure the second issue is due to modern forge using Forge auto renaming tool and 1.12 using special source for remapping. May be wrong though ofc

**Comment by Juuxel** [comment]
We typically read the remapping tool (at least for modern Forge) directly from the toolchain json, but that doesn't even apply here since the final named jar is always remapped using Tiny Remapper.

> Using MCP 1.15 (or older) mappings fails, as they seem to be provided as a .zip file (from Forge Maven, if I'm not mistaken), and not .jar

I don't think this is a Loom issue, you just need to make the dependency notation use zips (check gradle's own docs for that)

**Comment by MsRandom** [comment]
@SokyranTheDragon Just doing `"de.oceanlabs.mcp:mcp_stable:39-1.12@zip"` should be enough to tell gradle to select a zip artifact.

**Comment by MsRandom** [comment]
I'm not sure what that mappings issue is; I didn't get that with my testing, but I'll look into it as soon as I can.

**Comment by MsRandom** [comment]
> I'm 90% sure the second issue is due to modern forge using Forge auto renaming tool and 1.12 using special source for remapping. May be wrong though ofc 

Forge's auto rename tool isn't used at all when using a legacy version, there's a hard coded version of special source that's used (for renaming to srg after patching), which is exactly what ForgeGradle does for older versions.

**Comment by SokyranTheDragon** [comment]
> @SokyranTheDragon Just doing `"de.oceanlabs.mcp:mcp_stable:39-1.12@zip"` should be enough to tell gradle to select a zip artifact.

Huh, I didn't know it's possible to do that. I feel that if/once this gets merged in, there should be some note in the docs.

@Juuxel I wasn't trying to say it's an issue, and more trying to point out it may be something that'd be worth checking out since a lot of older mods still use MCP mappings. But yea... I'm not best at Gradle here, so I wasn't aware you could just putt the `@zip` part. Still, it wasn't 100% obvious why it was failing for until I started digging into it.


And for the remapping issue: I've tried clearing my cache and reloading the projects in IntelliJ, but it didn't change anything about it sadly. `getCollisionBoundingBox` is still mapped as `getItem`.

**Comment by SokyranTheDragon** [comment]
Ah, found the issue. The intermediary I was using was the culprit. However, I wasn't able to catch that, as even if I changed it the old one was still used - I had to remove the run configuration and let it be created again. Anyway, what gave it away is that I've noticed this trying to run the game:
`[2022-07-27 01:46:23] [INFO   ] Srg2Mcp: C:\Users\Sokyran\.gradle\caches\fabric-loom\1.12.2\de.oceanlabs.mcp.mcp_snapshot.1_12_2.20201028-1.16.3-forge-1.12.2-14.23.5.2860\mcp_to_srg.srg`

So anyway, solution: changing `setIntermediaryUrl("https://maven.legacyfabric.net/net/legacyfabric/intermediary/${minecraft_version}/intermediary-${minecraft_version}-v2.jar")` to `setIntermediaryUrl("https://maven.legacyfabric.net/net/legacyfabric/intermediary-old/${minecraft_version}/intermediary-old-${minecraft_version}-v2.jar")`

Followed by removing and re-generating the run configuration. Seems something was off with the non-old version? Anyway, I'm using legacyfabric intermediary since fabric ones seem to be 1.14 (1.13?) and up.


Anyway, right now I'm unable to run the game in the deofuscated environment. I'll be honest, it's way too late for me to deal with this right now. Way past midnight. 2:40 AM to be precise. So if anyone would be able to point me in the right direction, I'd appreciate it. Here's the crash: https://gist.github.com/SokyranTheDragon/1c177f7feb40549924393b3feaf783ba

**Comment by MsRandom** [comment]
That crash seems to suggest that the `mcp_to_srg` environment variable is not being used properly, not entirely sure why that would happen though.

**Comment by thecatcore** [comment]
You should avoid using the old intermediaries if possible as they are not used anymore in production at all.

**Comment by SokyranTheDragon** [comment]
@arthurbambou this one was the the only one that I got to work with no issues. With others it either failed for me, or I ended up with incorrectly named method names.

Anyway, for other stuff I've tried:

I've tried to see what changing `field_150915_c` to `EFFECTIVE_ON` using the debugger would do, and seems that it works if I did that. So yeah, seems something wrong.

I've then tried to clean all the caches I could, and seems that it somehow failed to set up MC/Forge? I did some digging in the cache and seems like `forge-1.12.2-14.23.5.2860-minecraft-merged-named.jar` and `forge-1.12.2-14.23.5.2860-minecraft-merged-named-sources.jar` failed to generate. Anyway, managed to fix that.

Still, even after trying to clean all up it failed. So I've decided to completely remove and reinstall Gradle and IntelliJ. Still the same issue. May try more stuff later/tomorrow. And in case anyone wants to check my stuff, here's my project (with a branch in which I've swapped to Architectury Loom): https://github.com/SokyranTheDragon/Minor-Integrations-and-Additions-MIA-/tree/architectury-loom

**Comment by wagyourtail** [comment]
I got it building for me by replacing the top of your build gradle with
```diff
@@ -1,8 +1,27 @@
+buildscript {
+    repositories {
+        maven {
+            name = "forge"
+            url = "https://files.minecraftforge.net/maven"
+        }
+        maven { url "https://maven.fabricmc.net/" }
+        maven { url "https://maven.architectury.dev/" }
+        maven { url 'https://repo.spongepowered.org/maven' }
+        maven { url 'https://jitpack.io' }
+    }
+    dependencies {
+        classpath 'com.github.MsRandom:architectury-loom:0.12.0-legacyfg3-SNAPSHOT'
+    }
+}
+
+
 plugins {
-    id "dev.architectury.loom" version "0.12.0.9999"
+    id 'java'
     id "com.matthewprenger.cursegradle" version "1.4.0"
 }
 
+apply plugin: "dev.architectury.loom"
+
 sourceCompatibility = targetCompatibility = JavaVersion.VERSION_1_8
 
 archivesBaseName = project.archives_base_name
@@ -10,8 +29,6 @@ version = "${project.minecraft_version}-${project.mod_version}"
 group = project.maven_group
 
 loom {
-    setIntermediaryUrl("https://maven.legacyfabric.net/net/legacyfabric/intermediary-old/${minecraft_version}/intermediary-old-${minecraft_version}-v2.jar")
-    customMinecraftManifest.set("https://meta.legacyfabric.net/v2/manifest/${minecraft_version}")
 
     // since loom 0.10, you are **required** to use the
     // "forge" block to configure forge-specific features,
     ```
     
     yes. the mappings are a bit messed up as you indicated, I expect thats because of yarn intermediary mistakes.
     
     I wouldn't use `-old` personally, I'd just deal with the few mis-mappings, they shouldn't be too common

**Comment by MsRandom** [comment]
I'd suggest moving the repositories(including jitpack) back to the settings.gradle file but with a pluginManagement block to allow gradle to find the plugin properly from there, that way the plugins dsl can still be used rather than an imperative plugin apply.

**Comment by SokyranTheDragon** [comment]
@wagyourtail I've tried changing what you pointed out. However, removing intermediary url (and clearing the cache, as it seems it's not re-downloaded unless you remove it) caused those errors:

```
Not Found for https://maven.fabricmc.net/net/fabricmc/intermediary/1.12.2/intermediary-1.12.2-v2.jar
```
and
```
A problem occurred configuring root project 'MIA'.
> Failed to setup Minecraft, java.lang.RuntimeException: java.io.UncheckedIOException: Failed to provide intermediate mappings
```

However, after I restored the intermediary, the other changes did not really do anything for me. Still getting the same exception as earlier.

**Comment by thecatcore** [comment]
That's an issue that also occures when switching from old to new legacy fabric intermediary.

**Comment by SokyranTheDragon** [comment]
Welp, I've tried to remap my source from MCP to LegacyFabric Yarn, but it ended up with quite a bit of conflicts when remapping dependencies.

I also though I may try building the jar and trying it outside of deobf environment. I'm assuming I need to run `remapJar`? As just the jar file from running `jar` in gradle was causing crashes in MC. If so, seems I can't do it as there's some conflicts?
```
> Task :remapJar FAILED
Mapping source name conflicts detected:
  net/minecraft/world/gen/ChunkGeneratorHell METHOD isInsideStructure ((Lnet/minecraft/world/World;Ljava/lang/String;Lnet/minecraft/util/math/BlockPos;)Z) -> [net/minecraft/world/gen/IChunkGenerator/func_193414_a, net/minecraft/world/gen/ChunkGeneratorHell/func_193413_a]
(Repeat the same conflict for a bunch of other classes)
```

Am I even using the correct intermediary? I'm using the LegacyFabric one for 1.12, is this the one I should use? Couldn't find another 1.12.2 one, and using ones from future versions (obviously) failed. I've also tried both the intermediary and intermediary-old, none allowed me to either run or remap the jar. But the non-old intermediary also had some issues with incorrect mapping, for example it would cause the remapping of `Block.getCollisionBoundingBox` as `Block.getItem`, or do the opposite (seemingly at random).

I'll be honest, I'm out of ideas to try now. Building the project is fine. Building the jar is fine as well. Running it or trying to remap the built jar is not.

Since the earlier crash log I posted doesn't have the full log, here it is: https://gist.github.com/SokyranTheDragon/1c33bd2a8ec52ed339739a222ea06448

Going to most likely try running it on a fresh project tomorrow. But outside of that, I honestly have no idea what else I could try to get it to work.

**Comment by SokyranTheDragon** [comment]
Okay, decided to start with a fresh setup to see if it changes anything. Used the 1.12.2 - 14.23.5.2860 as the base, but basically replaced all gradle stuff. Anyway, no change. Still the same issues, exceptions, etc.

Here's the repo if anyone's interested, it's pretty small: https://github.com/SokyranTheDragon/TestMod

Though, since I didn't have to worry about mapping conflicts with dependencies, I tried to run it with legacyfabric yarn. First of all, it was giving me "unfixable conflicts" error, until I ended up using old enough yarn mappings that it was happy. It still would not run though, here's the log: https://gist.github.com/SokyranTheDragon/4cf6e855ac530504d273edf802ef3cdf

When I did it, it refused to launch at first. It was complaining about modified .jar file. It said to use `-Dfml.ignoreInvalidMinecraftCertificates=true` if I really want to run, which I did.


Anyway, that's all from me. No idea what else I could even try at this point. Nothing I did seemed to work.

**Comment by MsRandom** [comment]
After a bit of testing, it appears there's a problem in `MappingsProviderImpl::getMcpToSrg`, it either doesn't add the class mappings, or it adds them twice. I'll investigate further to see why that's happening.

**Comment by MsRandom** [comment]
Should be fixed with commit ac887924. Legacy yarn mappings will still not work with legacy forge however, as it does some stuff that expects the class mappings to be in SRG (similarly to 1.19 where they assume they're officially mapped), and fixing that is outside of the scope of this PR.

**Pull request closed without merging**
